optaplanner-website-root/content/blog/2022/09/26/DynamicScalingOfOptaPlannerWorkloadOnOpenShift.adoc

```diff
@@ -0,0 +1,93 @@
+
+The https://github.com/kiegroup/optaplanner-quickstarts/tree/development/technology/kubernetes[Kubernetes demo],
+based on the https://www.optaplanner.org/learn/useCases/schoolTimetabling.html[school timetabling] use case,
+has been updated to show the dynamic scaling.
+
+== Architecture
+
+The architecture has evolved, replacing some existing components and adding a new one, too.
+
+image::demoArchitecture.svg[OpenShift demo architecture diagram]
+
+The OptaPlanner operator depends on two other operators:
+
+* https://artemiscloud.io/[ArtemisCloud] operator, which creates ActiveMQ Artemis broker and queues
+* https://keda.sh/[KEDA] for scaling the solver deployment, represented by the `school-timetabling` pods in the picture
+
+The _Demo App_ generates datasets, stores them in the PostgreSQL database, and sends a message to the `school-timetabling-problem` queue per each dataset.
+The _School Timetabling_ project, which defines the optimization problem, reads a single message from the `school-timetabling-problem` queue, loads the dataset
+from the database, and solves it. After that, it saves the solution back to the database and sends a message to the `school-timetabling-solution` queue to let the _Demo App_ know that the solution is ready for taking.
+
+=== ActiveMQ
+
+The use case for messaging in this architecture is a shared job queue. ActiveMQ Artemis is a good fit here as it provides the exactly-once delivery.
+If a solver pod dies, the ActiveMQ Artemis broker detects a broken connection and redelivers the message to any other available pod.
+
+==== Batch delivery
+
+By default, the ActiveMQ Artemis distributes messages in batches to reduce needless network round trips.
+While this is a favorable behavior for large amounts of messages requiring relatively short processing time, our case is different.
+
+We expect fewer messages that, on the other hand, take a long time to process, as it involves optimizing a dataset by OptaPlanner.
+Also, we have a different goal; instead of minimizing the latency, we aim to parallelize solving of multiple datasets as much as possible.
+Each active solver pod must be solving one dataset at any moment.
+
+Batch message delivery hinders scalability because if the first solver pod prefetches multiple messages, the other pods may sit idle.
+
+The message prefetch is switched off by using the client-side JMS connection parameter `jms.prefetchPolicy.queuePrefetch=0`.
+See the `quarkus.qpid-jms.url` property in https://github.com/kiegroup/optaplanner-quickstarts/blob/development/technology/kubernetes/school-timetabling/src/main/resources/application.properties[application.properties] of the _School Timetabling_ project.
+
+==== Atomicity
+
+Receiving the input problem message and sending the output message informing that solving has ended must happen in a transaction
```

Comment:
```suggestion
A transaction must receive the input problem message and send the output message informing that solving has ended.
```

Reply:
I might have changed the meaning so if I did please correct. I was trying to make the sentence more direct and less round about

Reply:
The "transaction" cannot be a subject here, but I will rearrange the sentence.